rt/transports/http/src/main/java/org/apache/cxf/transport/http/AbstractHTTPDestination.java

```diff
@@ -410,8 +410,12 @@ private boolean isWSAddressingReplyToSpecified(Exchange ex) {
             public Principal getUserPrincipal() {
                 try {
                     return req.getUserPrincipal();
-                } catch (Exception ex) {
-                    return null;
+                } catch (Exception e) {
```

Comment:
@rzo1 thanks for catching it, @ffang do you have an idea why `request` could be `null` in `ServletApiRequest`? (may be we have problem somewhere with Jetty 12 integration)

```
     at org.eclipse.jetty.server.Request.getAuthenticationState(Request.java:949)
	at org.eclipse.jetty.security.AuthenticationState.getAuthenticationState(AuthenticationState.java:45)
	at org.eclipse.jetty.ee10.servlet.ServletApiRequest.getAuthentication(ServletApiRequest.java:254)
	at org.eclipse.jetty.ee10.servlet.ServletApiRequest.getUndeferredAuthentication(ServletApiRequest.java:259)
	at org.eclipse.jetty.ee10.servlet.ServletApiRequest.getUserPrincipal(ServletApiRequest.java:478)
	at org.apache.cxf.transport.http.AbstractHTTPDestination$2.getUserPrincipal(AbstractHTTPDestination.java:412)
```

Reply:
In CXF 4.0.x where was'nt a a catch block at all. So this PR only cures a symptom and it might be worth to dig deeper on why it's actually null. Don't know enough about CXF internals though 😄

Reply:
I'm on it guys.

Freeman

Reply:
Hi Guys,

I think I figured out the root cause of the NPE. 

This will happen in case of oneway or ReplyTo is specified when ws-addressing is used, which means we need to switch thread context on the server side. Basically it's pause the current interceptor chain and resume it in another thread, and before the resumed current interceptor chain is completed, the underlying transport might discard any data on the original stream, that's why we need to cache the InputStream in this case and use it later if necessary. And this peace of code is in
org.apache.cxf.ws.addressing.impl.InternalContextUtils.rebaseResponse. Specifically we have
```
             if (ContextUtils.retrieveAsyncPostResponseDispatch(inMessage) && !robust) {
                        //need to suck in all the data from the input stream as
                        //the transport might discard any data on the stream when this
                        //thread unwinds or when the empty response is sent back
                        DelegatingInputStream in = inMessage.getContent(DelegatingInputStream.class);
                        if (in != null) {
                            in.cacheInput();
                        }
                    }
```
Which in turn will call into org.apache.cxf.transport.http.AbstractHTTPDestination.setupMessage
```
       DelegatingInputStream in = new DelegatingInputStream(req.getInputStream()) {
            public void cacheInput() {
                if (!cached && (exchange.isOneWay() || isWSAddressingReplyToSpecified(exchange))) {
                    //For one-ways and WS-Addressing invocations with ReplyTo address,
                    //we need to cache the values of the HttpServletRequest
                    //so they can be queried later for things like paths and schemes
                    //and such like that.
                    //Please note, exchange used to always get the "current" message
                    exchange.getInMessage().put(HTTP_REQUEST, new HttpServletRequestSnapshot(req));
                }
                if (req.getContentLengthLong() != 0) {
                    super.cacheInput();
                }
            }
            private boolean isWSAddressingReplyToSpecified(Exchange ex) {
                AddressingProperties map = ContextUtils.retrieveMAPs(ex.getInMessage(), false, false, false);
                return map != null && !ContextUtils.isGenericAddress(map.getReplyTo());
            }
        };
```

So in this case the HttpServletRequestSnapshot which is a wrapper of original HttpServletRequest is saved in InMessage, and we should always use pre-saved HTTP_REQUEST from InMessage to populate SecurityContext, as pre-saved HTTP_REQUEST from InMessage could be the original HttpServletRequest or the Cached HttpServletRequest when it's necessary. 

We don't see this change is necessary for Jetty 11 and before. 
Just send a PR https://github.com/apache/cxf/pull/1823 to address this for Jetty12.

Thanks and Best Regards
Freeman

Reply:
Nice, I will close my PR and do a test in TomEE with a local build of your PR tmrw 🙃

Reply:
Hi @rzo1 ,

PR merged to main branch, so probably the snapshot build will be available from Apache Snapshot maven repo by tomorrow.  

Cheers
Freeman

Reply:
Hi @ffang ,

TomEE is happy with your fix. CI is green again ;-)
Thx for it !

Gruß
Richard

Reply:
Awesome! Thanks for the update @rzo1 !

Reply:
@ffang Did some more tests. 4.1.0-SNAPSHOT only produces 3 JAX-RS TCK Failures (3.1.5) - quite promising :)

Reply:
Great news! Thanks again for the update @rzo1 ! Please raise jira ticket and append TCK failures there so that we can take a close look.

Best Regards
Freeman

Reply:
@ffang https://issues.apache.org/jira/browse/CXF-9005